
Status: An up-to-date (and more correct) version of this patch is pending in
        upstream (https://bugzilla.mozilla.org/show_bug.cgi?id=261649). This
        one might not even work in runtime, make sure it passes the testsuite
        before merging.

diff -ur mozilla.old/nsprpub/configure.in mozilla/nsprpub/configure.in
--- mozilla.old/nsprpub/configure.in	2004-04-17 00:28:02.000000000 +0200
+++ mozilla/nsprpub/configure.in	2004-07-22 17:21:04.000000000 +0200
@@ -511,9 +511,12 @@
     OS_RELEASE=
     OS_TEST="${target_cpu}"
     case "${target_os}" in
-        linux*)       OS_ARCH=Linux ;;
-        solaris*)     OS_ARCH=SunOS OS_RELEASE=5 ;;
-        mingw*)       OS_ARCH=WINNT ;;
+        linux*)        OS_ARCH=Linux ;;
+        solaris*)      OS_ARCH=SunOS OS_RELEASE=5 ;;
+        mingw*)        OS_ARCH=WINNT ;;
+        kfreebsd*-gnu) OS_ARCH=GNU/kFreeBSD ;;
+        knetbsd*-gnu)  OS_ARCH=GNU/kNetBSD ;;
+        gnu*)          OS_ARCH=GNU ;;
     esac
 else
     OS_ARCH=`uname -s | sed -e 's|/|_|g'`
@@ -919,6 +922,7 @@
     fi
     AC_DEFINE(XP_UNIX)
     AC_DEFINE(FREEBSD)
+    AC_DEFINE(KFREEBSD)
     AC_DEFINE(HAVE_BSD_FLOCK)
     AC_DEFINE(HAVE_SOCKLEN_T)
     CFLAGS="$CFLAGS $(DSO_CFLAGS) -ansi -Wall"
@@ -1174,7 +1178,7 @@
 	esac
     ;;
 
-*-linux*)
+*-linux* | *-gnu* | *-k*bsd*-gnu)
     if test -z "$USE_NSPR_THREADS"; then
         USE_PTHREADS=1
         IMPL_STRATEGY=_PTH
@@ -1185,7 +1189,13 @@
     AC_DEFINE(_SVID_SOURCE)
     AC_DEFINE(_LARGEFILE64_SOURCE)
     AC_DEFINE(HAVE_FCNTL_FILE_LOCKING)
-    AC_DEFINE(LINUX)
+    AC_DEFINE(GLIBC)
+    case "$target" in
+      *-linux*) AC_DEFINE(LINUX) ;;
+      *-kfreebsd*-gnu) AC_DEFINE(KFREEBSD) ;;
+      *-knetbsd*-gnu) AC_DEFINE(KNETBSD) ;;
+      *-gnu*) AC_DEFINE(HURD) ;;
+    esac
     CFLAGS="$CFLAGS -ansi -Wall"
     CXXFLAGS="$CXXFLAGS -ansi -Wall"
     MDCPUCFG_H=_linux.cfg
@@ -1408,6 +1418,7 @@
 *-netbsd*)
     AC_DEFINE(XP_UNIX)
     AC_DEFINE(NETBSD)
+    AC_DEFINE(KNETBSD)
     AC_DEFINE(HAVE_BSD_FLOCK)
     USE_NSPR_THREADS=1
     MDCPUCFG_H=_netbsd.cfg
@@ -2207,7 +2218,7 @@
             _PTHREAD_LDFLAGS=-pthread
         fi
         ;;
-    *-linux*)
+    *-linux* | *-gnu* | *-k*bsd*-gnu)
         AC_DEFINE(_REENTRANT)
         ;;
     esac
@@ -2291,7 +2302,7 @@
         fi
     fi
     ;;
-*-linux*)
+*-linux* | *-gnu* | *-k*bsd*-gnu)
     if test -n "$USE_NSPR_THREADS"; then
         AC_DEFINE(_PR_LOCAL_THREADS_ONLY)
     fi
diff -ur mozilla.old/nsprpub/pr/include/md/_linux.cfg mozilla/nsprpub/pr/include/md/_linux.cfg
--- mozilla.old/nsprpub/pr/include/md/_linux.cfg	2004-07-22 15:18:32.000000000 +0200
+++ mozilla/nsprpub/pr/include/md/_linux.cfg	2004-07-22 16:52:31.000000000 +0200
@@ -39,7 +39,7 @@
 #define XP_UNIX
 #endif
 
-#ifndef LINUX
+#if !defined(LINUX) && defined(__linux__)
 #define LINUX
 #endif
 
diff -ur mozilla.old/nsprpub/pr/include/md/_pth.h mozilla/nsprpub/pr/include/md/_pth.h
--- mozilla.old/nsprpub/pr/include/md/_pth.h	2003-09-16 22:44:04.000000000 +0200
+++ mozilla/nsprpub/pr/include/md/_pth.h	2004-07-22 17:12:36.000000000 +0200
@@ -136,7 +136,7 @@
 	(!memcmp(&(t), &pt_zero_tid, sizeof(pthread_t)))
 #define _PT_PTHREAD_COPY_THR_HANDLE(st, dt)   (dt) = (st)
 #elif defined(IRIX) || defined(OSF1) || defined(AIX) || defined(SOLARIS) \
-	|| defined(HPUX) || defined(LINUX) || defined(FREEBSD) \
+	|| defined(HPUX) || defined(GLIBC) || defined(FREEBSD) \
 	|| defined(NETBSD) || defined(OPENBSD) || defined(BSDI) \
 	|| defined(VMS) || defined(NTO) || defined(DARWIN) \
 	|| defined(UNIXWARE)
@@ -188,7 +188,7 @@
 /*
  * These platforms don't have sigtimedwait()
  */
-#if (defined(AIX) && !defined(AIX4_3_PLUS)) || defined(LINUX) \
+#if (defined(AIX) && !defined(AIX4_3_PLUS)) || defined(GLIBC) \
 	|| defined(FREEBSD) || defined(NETBSD) || defined(OPENBSD) \
 	|| defined(BSDI) || defined(VMS) || defined(UNIXWARE) \
 	|| defined(DARWIN)
@@ -228,7 +228,7 @@
 #define PT_PRIO_MAX            sched_get_priority_max(SCHED_OTHER)
 #endif /* defined(_PR_DCETHREADS) */
 
-#elif defined(LINUX) || defined(FREEBSD)
+#elif defined(GLIBC) || defined(FREEBSD)
 #define PT_PRIO_MIN            sched_get_priority_min(SCHED_OTHER)
 #define PT_PRIO_MAX            sched_get_priority_max(SCHED_OTHER)
 #elif defined(NTO)
@@ -283,7 +283,7 @@
 		onemillisec.tv_nsec = 1000000L;			\
         nanosleep(&onemillisec,NULL);			\
     PR_END_MACRO
-#elif defined(HPUX) || defined(LINUX) || defined(SOLARIS) \
+#elif defined(HPUX) || defined(GLIBC) || defined(SOLARIS) \
 	|| defined(FREEBSD) || defined(NETBSD) || defined(OPENBSD) \
 	|| defined(BSDI) || defined(NTO) || defined(DARWIN) \
 	|| defined(UNIXWARE)
diff -ur mozilla.old/nsprpub/pr/include/md/prosdep.h mozilla/nsprpub/pr/include/md/prosdep.h
--- mozilla.old/nsprpub/pr/include/md/prosdep.h	2001-10-26 06:54:01.000000000 +0200
+++ mozilla/nsprpub/pr/include/md/prosdep.h	2004-07-22 17:05:22.000000000 +0200
@@ -87,7 +87,7 @@
 #elif defined(IRIX)
 #include "md/_irix.h"
 
-#elif defined(LINUX)
+#elif defined(LINUX) || defined(__GNU__) || defined(__GLIBC__)
 #include "md/_linux.h"
 
 #elif defined(OSF1)
diff -ur mozilla.old/nsprpub/pr/src/md/unix/unix.c mozilla/nsprpub/pr/src/md/unix/unix.c
--- mozilla.old/nsprpub/pr/src/md/unix/unix.c	2002-12-12 01:19:57.000000000 +0100
+++ mozilla/nsprpub/pr/src/md/unix/unix.c	2004-07-22 17:16:44.000000000 +0200
@@ -65,7 +65,7 @@
  * PRInt32* pointer to a _PRSockLen_t* pointer.
  */
 #if defined(HAVE_SOCKLEN_T) \
-    || (defined(LINUX) && defined(__GLIBC__) && __GLIBC__ >= 2)
+    || (defined(GLIBC) && __GLIBC__ >= 2)
 #define _PRSockLen_t socklen_t
 #elif defined(IRIX) || defined(HPUX) || defined(OSF1) || defined(SOLARIS) \
     || defined(AIX4_1) || defined(LINUX) || defined(SONY) \
diff -ur mozilla.old/nsprpub/pr/src/md/unix/uxrng.c mozilla/nsprpub/pr/src/md/unix/uxrng.c
--- mozilla.old/nsprpub/pr/src/md/unix/uxrng.c	2002-06-14 05:22:14.000000000 +0200
+++ mozilla/nsprpub/pr/src/md/unix/uxrng.c	2004-07-22 17:18:48.000000000 +0200
@@ -135,7 +135,7 @@
     return 0;
 }
 
-#elif (defined(LINUX) || defined(FREEBSD) || defined(NETBSD) || defined(OPENBSD))
+#elif (defined(LINUX) || defined(KFREEBSD) || defined(KNETBSD) || defined(OPENBSD))
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
diff -ur mozilla.old/nsprpub/pr/src/misc/prnetdb.c mozilla/nsprpub/pr/src/misc/prnetdb.c
--- mozilla.old/nsprpub/pr/src/misc/prnetdb.c	2004-01-22 23:22:37.000000000 +0100
+++ mozilla/nsprpub/pr/src/misc/prnetdb.c	2004-07-22 17:26:01.000000000 +0200
@@ -105,7 +105,7 @@
 #define _PR_HAVE_GETPROTO_R_INT
 #endif
 
-#if (defined(LINUX) && defined(__GLIBC__) && __GLIBC__ >= 2)
+#if (defined(GLIBC) && __GLIBC__ >= 2)
 #define _PR_HAVE_GETPROTO_R
 #define _PR_HAVE_5_ARG_GETPROTO_R
 #endif
diff -ur mozilla.old/nsprpub/pr/src/pthreads/ptio.c mozilla/nsprpub/pr/src/pthreads/ptio.c
--- mozilla.old/nsprpub/pr/src/pthreads/ptio.c	2004-02-04 02:31:33.000000000 +0100
+++ mozilla/nsprpub/pr/src/pthreads/ptio.c	2004-07-22 17:34:56.000000000 +0200
@@ -203,7 +203,7 @@
 #if defined(SOLARIS)
 #define _PRSockOptVal_t char *
 #elif defined(IRIX) || defined(OSF1) || defined(AIX) || defined(HPUX) \
-    || defined(LINUX) || defined(FREEBSD) || defined(BSDI) || defined(VMS) \
+    || defined(LINUX) || defined(GLIBC) || defined(FREEBSD) || defined(BSDI) || defined(VMS) \
     || defined(NTO) || defined(OPENBSD) || defined(DARWIN) \
     || defined(UNIXWARE) || defined(NETBSD)
 #define _PRSockOptVal_t void *
@@ -217,7 +217,7 @@
 #define _PRSelectFdSetArg_t void *
 #elif defined(IRIX) || (defined(AIX) && !defined(AIX4_1)) \
     || defined(OSF1) || defined(SOLARIS) \
-    || defined(HPUX10_30) || defined(HPUX11) || defined(LINUX) \
+    || defined(HPUX10_30) || defined(HPUX11) || defined(LINUX) || defined(GLIBC) \
     || defined(FREEBSD) || defined(NETBSD) || defined(OPENBSD) \
     || defined(BSDI) || defined(VMS) || defined(NTO) || defined(DARWIN) \
     || defined(UNIXWARE)
@@ -301,7 +301,7 @@
  * most current systems.
  */
 #if defined(HAVE_SOCKLEN_T) \
-    || (defined(LINUX) && defined(__GLIBC__) && __GLIBC__ >= 2)
+    || (defined(GLIBC) && __GLIBC__ >= 2)
 typedef socklen_t pt_SockLen;
 #elif (defined(AIX) && !defined(AIX4_1)) \
     || defined(VMS)
@@ -357,14 +357,14 @@
     int nbytes_to_send;                     /* size of header and file */
 #endif  /* SOLARIS */
 
-#ifdef LINUX
+#if defined(LINUX) || defined(GLIBC)
     /*
      * For sendfile()
      */
     int in_fd;                              /* descriptor of file to send */
     off_t offset;
     size_t count;
-#endif  /* LINUX */
+#endif  /* LINUX || GLIBC */
  
     PRIntervalTime timeout;                 /* client (relative) timeout */
 
@@ -1109,7 +1109,7 @@
 }
 #endif  /* SOLARIS */
 
-#ifdef LINUX 
+#if defined(LINUX)
 static PRBool pt_linux_sendfile_cont(pt_Continuation *op, PRInt16 revents)
 {
     ssize_t rv;
@@ -2544,7 +2544,7 @@
 
 #endif  /* SOLARIS */
 
-#ifdef LINUX
+#if defined(LINUX)
 /*
  * pt_LinuxSendFile
  *
@@ -2983,7 +2983,7 @@
                 rv = setsockopt(
                     fd->secret->md.osfd, level, name,
                     (char*)&value, sizeof(PRIntn));
-#ifdef LINUX
+#if defined(LINUX)
                 /* for pt_LinuxSendFile */
                 if (name == TCP_NODELAY && rv == 0) {
                     fd->secret->md.tcp_nodelay = value;
@@ -3255,7 +3255,7 @@
 };
 
 #if defined(HPUX) || defined(OSF1) || defined(SOLARIS) || defined (IRIX) \
-    || defined(AIX) || defined(LINUX) || defined(FREEBSD) || defined(NETBSD) \
+    || defined(AIX) || defined(LINUX) || defined(GLIBC) || defined(FREEBSD) || defined(NETBSD) \
     || defined(OPENBSD) || defined(BSDI) || defined(VMS) || defined(NTO) \
     || defined(DARWIN) || defined(UNIXWARE)
 #define _PR_FCNTL_FLAGS O_NONBLOCK
@@ -4747,7 +4747,7 @@
 
 #include <sys/types.h>
 #include <sys/time.h>
-#if !defined(SUNOS4) && !defined(HPUX) && !defined(LINUX)
+#if !defined(SUNOS4) && !defined(HPUX) && !defined(LINUX) && !defined(GLIBC)
 #include <sys/select.h>
 #endif
 
diff -ur mozilla.old/nsprpub/config/nsinstall.c mozilla/nsprpub/config/nsinstall.c
--- mozilla.old/nsprpub/config/nsinstall.c	2002-12-12 01:29:08.000000000 +0100
+++ mozilla/nsprpub/config/nsinstall.c	2004-07-27 01:11:32.000000000 +0200
@@ -95,7 +95,7 @@
 }
 #endif /* NEXTSTEP */
 
-#ifdef LINUX
+#if defined(LINUX) || defined(GLIBC)
 #include <getopt.h>
 #endif
 
